pkg/controller/docdb/dbcluster/dbcluster.go

```diff
@@ -0,0 +1,304 @@
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+
+	meta.SetExternalName(cr, awsclient.StringValue(resp.DBCluster.DBClusterIdentifier))
+
+	return upd, nil
+}
+
+func (e *hooks) preCreate(ctx context.Context, cr *svcapitypes.DBCluster, obj *svcsdk.CreateDBClusterInput) error {
+	obj.AvailabilityZones = cr.Spec.ForProvider.AvailabilityZones
+	obj.BackupRetentionPeriod = cr.Spec.ForProvider.BackupRetentionPeriod
+	obj.DBClusterIdentifier = cr.Spec.ForProvider.DBClusterIdentifier
+	obj.DBClusterParameterGroupName = cr.Spec.ForProvider.DBClusterParameterGroupName
+	obj.DBSubnetGroupName = cr.Spec.ForProvider.DBSubnetGroupName
+	obj.DeletionProtection = cr.Spec.ForProvider.DeletionProtection
+	obj.EnableCloudwatchLogsExports = cr.Spec.ForProvider.EnableCloudwatchLogsExports
+	obj.Engine = cr.Spec.ForProvider.Engine
+	obj.EngineVersion = cr.Spec.ForProvider.EngineVersion
+	obj.KmsKeyId = cr.Spec.ForProvider.KMSKeyID
+
+	// TODO: Credentials from secret?
+	obj.MasterUserPassword = cr.Spec.ForProvider.MasterUserPassword
+	obj.MasterUsername = cr.Spec.ForProvider.MasterUsername
+
+	obj.Port = cr.Spec.ForProvider.Port
+	obj.PreSignedUrl = cr.Spec.ForProvider.PreSignedURL
+	obj.PreferredBackupWindow = cr.Spec.ForProvider.PreferredBackupWindow
+	obj.PreferredMaintenanceWindow = cr.Spec.ForProvider.PreferredMaintenanceWindow
+	obj.StorageEncrypted = cr.Spec.ForProvider.StorageEncrypted
+	obj.Tags = generateTags(cr.Spec.ForProvider.Tags)
+	obj.VpcSecurityGroupIds = cr.Spec.ForProvider.VPCSecurityGroupIDs
+
+	return nil
+}
+
+func (e *hooks) postCreate(ctx context.Context, cr *svcapitypes.DBCluster, resp *svcsdk.CreateDBClusterOutput, upd managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	meta.SetExternalName(cr, awsclient.StringValue(resp.DBCluster.DBClusterIdentifier))
+
+	return upd, nil
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.DBCluster, obj *svcsdk.DeleteDBClusterInput) (bool, error) {
+	// Skip if cluster is already in deleting state
+	if awsclient.StringValue(cr.Status.AtProvider.Status) == svcapitypes.DocDBInstanceStateDeleting {
+		return true, nil
+	}
+
+	obj.DBClusterIdentifier = awsclient.String(meta.GetExternalName(cr))
+	obj.SkipFinalSnapshot = awsclient.Bool(true)
+	return false, nil
+}
+
+func filterList(cr *svcapitypes.DBCluster, list *svcsdk.DescribeDBClustersOutput) *svcsdk.DescribeDBClustersOutput {
+	filtered := []*svcsdk.DBCluster{}
+
+	id := meta.GetExternalName(cr)
+
+	for _, cluster := range list.DBClusters {
+		if awsclient.StringValue(cluster.DBClusterIdentifier) == id {
+			filtered = append(filtered, cluster)
+		}
+	}
+
+	return &svcsdk.DescribeDBClustersOutput{
+		Marker:     list.Marker,
+		DBClusters: filtered,
+	}
+}
+
+// TODO(knappek): split this out as it is used in several controllers
+func (t *tagger) Initialize(ctx context.Context, mg resource.Managed) error {
```

Comment:
We have this implementation in some resources but it causes some issues since it reorders the tags as described [here](https://github.com/crossplane/provider-aws/issues/566#issuecomment-795314807).

Reply:
You are right. I didn't notice it because I haven't added unit tests for `tagger`.

I replaced it with another implementation that only uses the map to check if the key already exists and append it to the end of the list if otherwise. I also wrote a wrapper around `resource.GetExternalTags` that will return a sorted list instead of a map.